working/0731 - horizontal inference/feature-specification.md

```diff
@@ -0,0 +1,485 @@
+  _\<initializerListEntry\>_, _\<metadatum\>_, _\<newExpression\>_,
+  _\<constObjectExpression\>_, and _\<constructorInvocation\>_.
+
+- A "non-constructor invocation" is any other invocation.  For non-constructor
+  invocations, the target always takes the syntactic form of an expression,
+  though it may not actually have expression semantics (e.g. in the case where
+  the target refers to a top level function or static method).
+
+- The "target function type" of an invocation is the type of the invocation's
+  target, if it is a function type.  For constructor invocations, this is the
+  function type of the corresponding constructor.  For non-constructor
+  invocations, the target function type is determined as part of type inference,
+  as specified below.
+
+- A "dynamic invocation" is a non-constructor invocation that has no target
+  function type.
+
+- "Type inference" is a phase of compilation in which a static type is assigned
+  to each expression in a Dart program.  It is defined in the [Top-level and
+  local type inference][inference] document.
+
+- The "type inference algorithm" is a recursive process of visiting all the
+  statements and expressions in the program in a well-defined order, with the
+  nesting structure of the recursion reflecting the nesting structure of
+  statements and expressions in the program's syntax tree.  The precise order in
+  which statements and expressions are visited is defined in the [flow
+  analysis][] document.  Note that this document is unfinished; one particular
+  piece of information missing from it is the precise order in which arguments
+  to an invocation are visited.  This proposal supplies that information.
+
+- To "perform inference" on a given expression or statement means to execute the
+  portion of the type inference algorithm that visits that expression or
+  statement.
+
+- A "type schema" (defined [here][type schemas]) is a generalization of the
+  normal Dart type syntax, extended with a type known as "the unknown type"
+  (denoted `_`), which allows representation of incomplete type information.
+
+- An expression's "type context" is a type schema representing what is known
+  about the probable type of the expression just before the portion of the type
+  inference algorithm that visits it.  This, together with other information
+  defined in [flow analysis][], constitutes the _input_ to the corresponding
+  type inference step.
+
+- An "implicit type argument" is a type argument to a generic invocation whose
+  precise value is not directly specified in code, but is instead determined
+  automatically by the compiler during type inference.
+
+- A "type constraint" (defined [here][type constraints]) is a limitation on the
+  possible values that may be assigned to an implicit type argument.
+
+- "Subtype constraint generation" (defined [here][subtype constraint
+  generation]) is an operation on two type schemas that produces a set of type
```

Comment:
I see that your suggested rephrasing is more precise, but personally, I find it harder to grasp quickly.  Given that this is meant to be a reference to a definition in another doc, rather than a normative definition in and of itself, my preference is to leave it as is.